projects/RabbitMQ.Client/client/impl/SocketFrameHandler.cs

```diff
@@ -238,45 +238,23 @@ public void Close()
             }
         }
 
-        public InboundFrame ReadFrame()
+        public ValueTask<InboundFrame> ReadFrameAsync()
         {
-            return InboundFrame.ReadFrom(_reader, _frameHeaderBuffer, _amqpTcpEndpoint.MaxMessageSize);
+            return InboundFrame.ReadFromPipeAsync(_pipeReader, _amqpTcpEndpoint.MaxMessageSize);
         }
 
-        public void SendHeader()
+        public bool TryReadFrame(out InboundFrame frame)
         {
-#if NETSTANDARD
-            var headerBytes = new byte[8];
-#else
-            Span<byte> headerBytes = stackalloc byte[8];
-#endif
-
-            headerBytes[0] = (byte)'A';
-            headerBytes[1] = (byte)'M';
-            headerBytes[2] = (byte)'Q';
-            headerBytes[3] = (byte)'P';
-
-            if (Endpoint.Protocol.Revision != 0)
-            {
-                headerBytes[4] = 0;
-                headerBytes[5] = (byte)Endpoint.Protocol.MajorVersion;
-                headerBytes[6] = (byte)Endpoint.Protocol.MinorVersion;
-                headerBytes[7] = (byte)Endpoint.Protocol.Revision;
-            }
-            else
-            {
-                headerBytes[4] = 1;
-                headerBytes[5] = 1;
-                headerBytes[6] = (byte)Endpoint.Protocol.MajorVersion;
-                headerBytes[7] = (byte)Endpoint.Protocol.MinorVersion;
-            }
+            return InboundFrame.TryReadFrameFromPipe(_pipeReader, _amqpTcpEndpoint.MaxMessageSize, out frame);
+        }
 
-#if NETSTANDARD
-            _writer.Write(headerBytes, 0, 8);
+        public void SendHeader()
+        {
+#if NET6_0_OR_GREATER
+            _pipeWriter.AsStream().Write(ProtocolHeader);
```

Comment:
Does the stream need to be managed that is returned by `AsStream`?

Reply:
I don't know why `.AsStream()` is necessary here, because `_pipeWriter.Write()` suffices. In addition, no preprocessor directive is necessary.

Reply:
I was more wondering if you would need to dispose the stream returned by `AsStream`. It seems though when you dispose it, it does complete the underlying writer

https://github.com/dotnet/runtime/issues/28965
https://github.com/dotnet/corefx/pull/36160

It seems though the PipeStream doesn't necessarily need to be disposed. So maybe adding a comment here would help to make it clear the stream is deliberately not disposed or you could use AsStream and set leaveOpen to true and then dispose the stream when done.

Reply:
It doesn't need to be disposed AFAICT, it mostly provides convenience methods to more easily adapt Stream based APIs to Pipelines.

Reply:
Won't the GC eventually call the dispose method causing the writer to be completed?

Reply:
My input was about making things explicit with a comment or switching to `leaveOpen` to `true` and dispose the stream. Then you never have to ask that question to yourself again when you look at this code a few months down the line

Reply:
I agree it is more straightforward, I asked because @lukebakken marked this as resolved.
Made me wonder if I am missing something.

Reply:
I can add a comment!